superduperdb/components/model.py

```diff
@@ -429,14 +430,35 @@ def __call__(self, r):
         return args, kwargs
 
 
+def init_decorator(func):
+    """Decorator to set _is_initialized to True after init method is called.
+
+    :param func: init function.
+    """
+
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+        result = func(self, *args, **kwargs)
+        self._is_initialized = True
+        return result
+
+    return wrapper
+
+
 class ModelMeta(LeafMeta):
     """Metaclass for the `Model` class and descendants # noqa."""
 
     def __new__(mcls, name, bases, dct):
         """Create a new class with merged docstrings # noqa."""
+        # Ensure the instance is initialized before calling predict/predict_batches
+        if 'predict' in dct:
+            dct['predict'] = ensure_initialized(dct['predict'])
+        if 'predict_batches' in dct:
+            dct['predict_batches'] = ensure_initialized(dct['predict_batches'])
+        # If instance call init method, set _is_initialized to True
+        if 'init' in dct:
+            dct['init'] = init_decorator(dct['init'])
         cls = super().__new__(mcls, name, bases, dct)
-        cls.predict_batches = ensure_initialized(cls.predict_batches)
-        cls.predict = ensure_initialized(cls.predict)
```

Comment:
Decoration should not be done at the class level, we decorate instance methods.

Reply:
I see, thanks.